File: java/client/src/main/java/org/apache/rocketmq/client/java/impl/producer/ProducerSettings.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.rocketmq.client.java.impl.producer;

import apache.rocketmq.v2.Publishing;
import apache.rocketmq.v2.Resource;
import apache.rocketmq.v2.Settings;
import com.google.common.base.MoreObjects;
import com.google.common.util.concurrent.Futures;
import com.google.protobuf.util.Durations;
import java.time.Duration;
import java.util.Set;
import java.util.stream.Collectors;
import org.apache.rocketmq.client.java.impl.ClientSettings;
import org.apache.rocketmq.client.java.impl.ClientType;
import org.apache.rocketmq.client.java.impl.UserAgent;
import org.apache.rocketmq.client.java.retry.ExponentialBackoffRetryPolicy;
import org.apache.rocketmq.client.java.retry.RetryPolicy;
import org.apache.rocketmq.client.java.route.Endpoints;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class ProducerSettings extends ClientSettings {
    private static final Logger LOGGER = LoggerFactory.getLogger(ProducerSettings.class);

    private final Set<String> topics;
    /**
     * If message body size exceeds the threshold, it would be compressed for convenience of transport.
     */
    private volatile int maxBodySizeBytes = 4 * 1024 * 1024;
    private volatile boolean validateMessageType = true;

    public ProducerSettings(String clientId, Endpoints accessPoint, ExponentialBackoffRetryPolicy retryPolicy,
        Duration requestTimeout, Set<String> topics) {
        super(clientId, ClientType.PRODUCER, accessPoint, retryPolicy, requestTimeout);
        this.topics = topics;
    }

    public int getMaxBodySizeBytes() {
        return maxBodySizeBytes;
    }

    public boolean isValidateMessageType() {
        return validateMessageType;
    }

    @Override
    public Settings toProtobuf() {
        final Publishing publishing = Publishing.newBuilder()
            .addAllTopics(topics.stream().map(name -> Resource.newBuilder().setName(name).build())
                .collect(Collectors.toList())).setValidateMessageType(validateMessageType).build();
        final Settings.Builder builder = Settings.newBuilder()
            .setAccessPoint(accessPoint.toProtobuf()).setClientType(clientType.toProtobuf())
            .setRequestTimeout(Durations.fromNanos(requestTimeout.toNanos())).setPublishing(publishing);
        return builder.setBackoffPolicy(retryPolicy.toProtobuf()).setUserAgent(UserAgent.INSTANCE.toProtoBuf()).build();
    }

    @Override
    public void applySettingsCommand(Settings settings) {
        final Settings.PubSubCase pubSubCase = settings.getPubSubCase();
        if (!Settings.PubSubCase.PUBLISHING.equals(pubSubCase)) {
            LOGGER.error("[Bug] Issued settings not match with the client type, clientId={}, pubSubCase={}, "
                + "clientType={}", clientId, pubSubCase, clientType);
            return;
        }
        final apache.rocketmq.v2.RetryPolicy backoffPolicy = settings.getBackoffPolicy();
        final Publishing publishing = settings.getPublishing();
        RetryPolicy exist = retryPolicy;
        this.retryPolicy = exist.inheritBackoff(backoffPolicy);
        this.validateMessageType = settings.getPublishing().getValidateMessageType();
        this.maxBodySizeBytes = publishing.getMaxBodySize();
        this.arrivedFuture.setFuture(Futures.immediateVoidFuture());
    }

    @Override
    public String toString() {
        return MoreObjects.toStringHelper(this)
            .add("clientId", clientId)
            .add("clientType", clientType)
            .add("accessPoint", accessPoint)
            .add("retryPolicy", retryPolicy)
            .add("requestTimeout", requestTimeout)
            .add("topics", topics)
            .add("maxBodySizeBytes", maxBodySizeBytes)
            .toString();
    }
}

File: php/ProducerSettings.php
<?php
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Apache\Rocketmq;

/**
 * ProducerSettings - Immutable configuration holder for Producer.
 *
 * Extracted from Producer constructor to encapsulate all configuration:
 * - Connection: endpoints, credentials, TLS
 * - Retry: maxAttempts, retryPolicy
 * - Timeouts: requestTimeout
 * - Topics: pre-warmed topic list
 * - Namespace: resource scoping
 *
 * Retry policy is mutable (updated by server settings via applyServerBackoffPolicy).
 */
class ProducerSettings
{
    private readonly string $clientId;
    private readonly int $maxAttempts;
    private readonly int $requestTimeout;
    private readonly array $topics;
    private readonly string $namespace;
    private readonly ?SessionCredentials $credentials;
    private readonly ?TlsCredentials $tlsCredentials;
    private readonly bool $sslEnabled;
    private ExponentialBackoffRetryPolicy|CustomizedBackoffRetryPolicy $retryPolicy;

    /**
     * @param string $endpoints gRPC server endpoint
     * @param array $options Configuration options
     */
    public function __construct(
        private readonly string $endpoints,
        array $options = []
    ) {
        if (empty($endpoints)) {
            throw new \InvalidArgumentException("Endpoints cannot be empty");
        }

        $this->clientId = $options['clientId'] ?? ('php-producer-' . getmypid() . '-' . time());
        $this->maxAttempts = $options['maxAttempts'] ?? 3;
        $this->requestTimeout = $options['requestTimeout'] ?? 3000;
        $this->topics = $options['topics'] ?? [];
        $this->namespace = $options['namespace'] ?? '';
        $this->tlsCredentials = $options['tlsCredentials'] ?? null;
        $this->sslEnabled = $options['sslEnabled'] ?? true;

        if (isset($options['credentials']) && $options['credentials'] instanceof SessionCredentials) {
            $this->credentials = $options['credentials'];
        } else {
            $this->credentials = null;
        }

        $this->retryPolicy = new ExponentialBackoffRetryPolicy($this->maxAttempts, 1000, 30000, 2.0);
    }

    /**
     * Apply server-side backoff policy settings.
     *
     * @param object $settings Server settings protobuf object
     * @param Logger $logger Logger for diagnostics
     * @return void
     */
    public function applyServerBackoffPolicy(object $settings, Logger $logger): void
    {
        if (!$settings->hasBackoffPolicy()) {
            return;
        }

        $serverPolicy = $settings->getBackoffPolicy();
        $logger->info("Received backoff policy from server");

        if ($serverPolicy->hasCustomizedBackoff()) {
            $customizedBackoff = $serverPolicy->getCustomizedBackoff();
            if ($customizedBackoff !== null && !ProtobufUtil::isRepeatedFieldEmpty($customizedBackoff->getNext())) {
                $this->retryPolicy = CustomizedBackoffRetryPolicy::fromProtobuf($serverPolicy);
                $logger->info("Updated retry policy from server backoff");
            }
        }
    }

    // ==================== Getters ====================

    public function getEndpoints(): string { return $this->endpoints; }
    public function getClientId(): string { return $this->clientId; }
    public function getMaxAttempts(): int { return $this->maxAttempts; }
    public function getRequestTimeout(): int { return $this->requestTimeout; }
    public function getTopics(): array { return $this->topics; }
    public function getNamespace(): string { return $this->namespace; }
    public function getCredentials(): ?SessionCredentials { return $this->credentials; }
    public function getTlsCredentials(): ?TlsCredentials { return $this->tlsCredentials; }
    public function isSslEnabled(): bool { return $this->sslEnabled; }

    /**
     * Get the current retry policy.
     *
     * @return ExponentialBackoffRetryPolicy|CustomizedBackoffRetryPolicy
     */
    public function getRetryPolicy(): ExponentialBackoffRetryPolicy|CustomizedBackoffRetryPolicy
    {
        return $this->retryPolicy;
    }

    /**
     * Replace the retry policy (used by server settings update).
     *
     * @param ExponentialBackoffRetryPolicy|CustomizedBackoffRetryPolicy $policy
     */
    public function setRetryPolicy(ExponentialBackoffRetryPolicy|CustomizedBackoffRetryPolicy $policy): void
    {
        $this->retryPolicy = $policy;
    }
}
